File: 백준/Silver/1012. 유기농 배추/유기농 배추.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {

    static int w,h;
    static int[] dx = {0,0,1,-1},dy={1,-1,0,0};
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int tc = Integer.parseInt(br.readLine());

        while (tc-->0){

            int[] input = Arrays.stream(br.readLine().split(" "))
                     .mapToInt(Integer::parseInt).toArray();
            h = input[0];
            w = input[1];
            int[][] map = new int[h][w];

            LinkedList<int[]> q = new LinkedList<>();

            for(int i=0;i<input[2];i++){
                int[] point = Arrays.stream(br.readLine().split(" "))
                         .mapToInt(Integer::parseInt).toArray();
                map[point[0]][point[1]] = 1;
                q.add(new int[]{point[1],point[0]});
            }


            int num = 1;
            while (!q.isEmpty()){

                int[] cur = q.poll();

                if(map[cur[1]][cur[0]] == 1){
                    num++;
                    map[cur[1]][cur[0]] = num;
                    dfs(cur,map,num);
                }
            }
            System.out.println(num-1);
        }

    }

    private static void dfs(int[] cur,int[][] map,int num) {

        for(int i=0;i<dx.length;i++){
            int nextX = cur[0]+dx[i];
            int nextY = cur[1]+dy[i];

            if(isRange(nextX,nextY) && map[nextY][nextX] == 1){
                map[nextY][nextX] = num;
                dfs(new int[]{nextX,nextY},map,num);
            }
        }
    }

    private static boolean isRange(int x, int y) {
        return x>=0 && y>=0 && x<w && y<h;
    }
}

File: 백준/Silver/1012. 유기농 배추/유기농 배추.kt

lateinit var map: Array<IntArray>
lateinit var visit: Array<BooleanArray>

fun main(args: Array<String>) {

    var tc = readln().toInt()

    while (tc --> 0){

        val (w, h, k) = readln().split(" ").map { it.toInt() }

        map = Array(h) { IntArray(w) }
        visit = Array(h) {BooleanArray(w)}

        for (i in 1..k) {
            val (x, y) = readln().split(" ").map { it.toInt() }
            map[y][x] = 1
        }

        var result: Int = 0

        for(i in 0..< h){
            for(j in 0..< w){

                if(!visit[i][j] && map[i][j] == 1){
                    dfs(j,i)
                    result++;
                }
            }
        }

        println(result)
    }
}

fun dfs(x: Int, y: Int){

    val dx = arrayOf(0,0,1,-1)
    val dy = arrayOf(1,-1,0,0)

    visit[y][x] = true

    for(i in dx.indices){

        val nextX = x+dx[i]
        val nextY = y+dy[i]

        if(isRange(nextX, nextY) && !visit[nextY][nextX] && map[nextY][nextX] == 1){
            dfs(nextX, nextY);
        }
    }
}

fun isRange(x:Int, y:Int): Boolean{
    return x>=0 && y>=0 && x<map[0].size && y< map.size;
}
